### DISCUSSION TITLE
Control inputs ignored in c++

### DISCUSSION THREAD

**Discussion by vranki** [JSBSim-Team/jsbsim#773] [Q&A]
Hi,

I'm trying to use JSBSim in my own simulator application written in c++.

The problem is that control inputs (stick & rudder) are ignored. I am setting fcs/*-cmd-norm values but it looks like the controls always stay centered. Throttle (fcs/throttle-cmd-norm) works as expected though. 
My code finds the property nodes for the values and uses setDoubleValue() to set the normalized value. I'm setting the values on each loop before FGFDMExec.Run() is called (not sure if it's necessary).

I am testing with f16 model, but the controls stay centered with any other models I've tried.

Any ideas how to debug this? I'm using latest release 1.1.12 on Ubuntu 20.04.

**Comment by seanmcleod** [comment]
@vranki I have a C++ based simulator I developed using JSBSim for test pilot training, acting as a VSS simulator.

So, in `JSBSim.cpp` there are calls to `copy_to_JSBSim()` and `copy_from_JSBSim()`.

```c++
        current_seconds = getcurrentseconds();                      // Seconds since 1 Jan 1970
        actual_elapsed_time = current_seconds - initial_seconds;    // Real world elapsed seconds since start
        sim_lag_time = actual_elapsed_time - FDMExec->GetSimTime(); // How far behind sim-time is from actual
                                                                    // elapsed time.

        // Copy control inputs in
        copy_to_JSBSim();

        for (int i=0; i<(int)(sim_lag_time/frame_duration); i++) {  // catch up sim time to actual elapsed time.
          result = FDMExec->Run();
          cycle_duration = getcurrentseconds() - current_seconds;   // Calculate cycle duration
          current_seconds = getcurrentseconds();                    // Get new current_seconds
          if (FDMExec->Holding()) break;
        }

        // Copy JSBSim model data out
        copy_from_JSBSim();

        if (play_nice) sim_nsleep(sleep_nseconds);
```

I added the following code to `copy_to_JSBSim()` to grab the inputs from the physical joystick, either a DirectInput based one or from Stirling Dynamics or Brunner and then pass the inputs to JSBSim as shown.

```c++
void copy_to_JSBSim()
{
.......
    // DirectInput inputs
    g_Joystick.Poll();

    if (g_Joystick.JoystickPresent())
    {
      FCS->SetDeCmd(g_Joystick.Pitch);
      FCS->SetDaCmd(g_Joystick.Roll);
    }
    FCS->SetDrCmd(DeadBand(g_Joystick.Rudder, 0.02));
    FCS->SetLBrake(g_Joystick.LeftBrake);
    FCS->SetRBrake(g_Joystick.RightBrake);

    // Stirling Inceptor inputs
    g_StirlingJoystick.Poll();
.........
}
```

**Reply by vranki** [reply]
Thanks, I didn't know about FCS class. I changed my code to get the FCS instance from FDMExec and set controls using SetD*Cmd() functions. But this changes nothing - controls are still centered. I tried values in ranges [-1,1] and [-100,100] but no change. FGFCS.h documentation talks about percents in setters and -1,1 in getters so I'm not sure which is correct.

My main loops looks quite the same except it calls Run() at fixed 60hz.

**Comment by seanmcleod** [comment]
Set a breakpoint in `SetDeCmd()` and see who is resetting the control commands.

**Reply by vranki** [reply]
I did that, but it's only called from my code. I also checked that after Run() the control values are correct (GetDeCmd() returns valid value). 

Could you confirm if the valid range is -1 to 1 or something else?

**Comment by seanmcleod** [comment]
> ...but it **looks like** the controls always stay centered.

You didn't mention how exactly you're determining this. Are you logging some properties, or are you flying the model connected to an external view and/or instruments and it doesn't appear to respond to joystick inputs. You mentioned the throttle working.

Are you possibly in trim mode all the time?

The range is typically `[-1, +1]` although I don't think there is anything that actually requires or enforces that other than convention.

For example look at my recent comment on another discussion today.

https://github.com/JSBSim-Team/jsbsim/discussions/771#discussioncomment-4252504

So at the end of the day what counts is the property used by the FDM developer for the functions that calculate moments and forces. They typically, but again no-one forces them to use an angular deflection, typically because that's how most aeronautical data is provided.

And so again typically within the FCS section there will be sections that convert a normalised `[-1, +1]` command to an angular command. But an FDM developer could use a completely different range for their command value and say a gain function to scale that to an angular deflection. Or even assume the command value is an angular deflection.

But given the convention I'm sure 99.99% of FDM models work on the assumption of a command range of `[-1, +1]`. Otherwise it would make life difficult for applications like FlightGear etc.

**Reply by vranki** [reply]
I have a CIGI visual displaying out of window view and instruments displaying the aircraft attitude. Also I'm looking for some control position properties such as fcs/elevator-pos-deg which stays at zero. If the aircraft is started from air, it just stalls and falls down, any control inputs have no effect on the way down.

My code does the following:
* setup directories & paths
* load aircraft model with LoadModel()
* get initial conditions pointer and load initial condition file with ic->Load()
* call RunIC()
* call SetDt()
* start engine with SetPropertyValue("propulsion/engine/set-running", true)

Flight loop does:
* get fcs pointer with GetFCS()
* set control positions with fcs->SetD*Cmd() (I verified the set values look correct)
* call Run()

I haven't found full example code that implements such simulation so this may be incorrect.

**Comment by seanmcleod** [comment]
One option to try is to use `JSBSim.cpp` and simply add in the call to `copy_to_JSBSim()` and `copy_from_JSBSim()` as I did above and see if that works.

Also confirm that the initial conditions make sense and that they are being loaded and set.

**Comment by vranki** [comment]
Hi, I'm still struggling with this. But there is some progress: I created a practically identical example app with just main() function with the same code I've been using. To my surprise that worked as expected and the controls move.

My original program that uses JSBSim is built as a Qt plugin. Could this cause JSBSim not to work as expected? It's using -std=gnu++11 -D_REENTRANT and -shared flags among others to build. This is getting above my understanding of c++ internals so I might be wrong.

I could try porting my app to be a standalone process instead of plugin to see if that would help. Anyway it's really weird behavior.

**Reply by bcoconni** [reply]
It is extremely unlikely that the problem you are facing is related to a compiler flag. The rule of thumb is that as long as the compilation succeeds and the code is able to (somehow) run then compiler settings are out of concern.

If you really need to change `fcs/elevator-pos-deg` then you should call the C++ method `FGFCS::SetDePos` instead of calling `FGFCS::SetDeCmd` which updates the property `fcs/elevator-cmd-norm`. The same goes for other `-pos` properties. The list of to "convert" FCS property names to C++ methods can be extracted from the bottom of the file `src/models/FGFCS.cpp`:
https://github.com/JSBSim-Team/jsbsim/blob/a919d4dd6fd7ba696d21e2c7f25cb35ceb670982/src/models/FGFCS.cpp#L717-L759

**Comment by bcoconni** [comment]
@vranki
Generally speaking, JSBSim manages a pair of property `-cmd-norm` and `-pos` for each control. The first property is meant to be the position that the control should target while the second property is the actual physical position of the control. There is no reason why the 2 values should match at any time : for example, there might be some lag between the `-cmd` setting and the actual control `-pos` change.

One typical example is for retracting gears where the landing gear only has two command positions "UP" and "DOWN" (`gear/gear-cmd-norm` can be set to either 0 or 1) while the gear might be located at an intermediate position (`gear/gear-pos-norm`):
```xml
        <kinematic name="Gear Control">
            <input>gear/gear-cmd-norm</input>
            <traverse>
                <setting>
                    <position>0</position>
                    <time>0</time>
                </setting>
                <setting>
                    <position>1</position>
                    <time>5</time>
                </setting>
            </traverse>
            <output>gear/gear-pos-norm</output>
        </kinematic>
```
It is the XML model that sets the laws that drive the `-pos` property from the value set by the `-cmd` property (i.e. target). If your model lack these laws then setting the `-cmd` properties has no effect. Most likely this is what happens in your case. However, as the controls are driven by C++ code in your case then it might make sense to shortcut the `-cmd` properties mechanism and set commands directly to `-pos` properties via the `FGFCS::SetD*Pos()` methods.

**Comment by vranki** [comment]
During the christmas vacations i found a solution but not the cause for the issue.

The issue is triggered when the process creates a  QCoreApplication in main(). QCoreApplication sets the process locale according to system settings. See full documentation at https://doc.qt.io/qt-6/qcoreapplication.html#locale-settings

As suggested, calling setlocale(LC_NUMERIC,"C") fixes the symptoms.

I'm not sure exactly what is the cause. Does jsbsim use some functions that depend on locale when reading or converting some data? Maybe when reading files it uses locale-aware functions which fail to parse strings?

In my system LC_NUMERIC is fi_FI.UTF-8 and for example locale-aware decimal separator is , instead of .

**Reply by bcoconni** [reply]
Indeed JSBSim uses locale dependent functions such as `atof()` which converts strings to numbers. According to StackOverflow [Locale-independent "atof"?](https://stackoverflow.com/questions/1333451/locale-independent-atof), `atof` is locale dependent.

**Comment by seanmcleod** [comment]
Hmm, I've run into locale specific issues with other software I've developed in the past, and in particular to do with data persistence and the different decimal separators when floating point values are persisted as strings.

I'm guessing the control input issue you were seeing then was really the tip of the iceberg potentially given all the floating point values that are read in from the FDM, unless the JSBSim routines for reading them ignores the locale setting, although there must be at least some code which makes use of the defined locale which was causing your issue.

Having a quick glance at `setlocale` for the Microsoft c-runtime.

https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170

> At program startup, the equivalent of the following statement is executed:
>
> setlocale( LC_ALL, "C" );

**Comment by bcoconni** [comment]
@vranki
By the way, has your issue been fixed regarding `fcs/elevator-pos-deg` ?

**Reply by vranki** [reply]
Yes, probably the whole simulation was broken due to invalid data read. I am now able to control the aircraft as expected.

**Comment by bcoconni** [comment]
FYI I have opened the issue #795 as JSBSim should at the very least report an error when the decimal separator in the flight model data does not match the locale.

**Reply by vranki** [reply]
Thanks! This should resolve this question. At least we now have a workaround for it.
